File: companion-apps/android/app/src/main/java/app/navon/bike/integration/cycling/OsmCyclingRoutingAdapter.kt
package app.navon.bike.integration.cycling

import java.net.HttpURLConnection
import java.net.URL
import java.util.Locale
import java.util.UUID
import kotlin.math.abs
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import app.navon.bike.domain.ActiveRouteSession
import app.navon.bike.domain.CoordinatePoint
import app.navon.bike.domain.NormalizedRoutePackage
import app.navon.bike.domain.RouteAlternative
import app.navon.bike.domain.RouteManeuver
import app.navon.bike.domain.RouteManeuverType
import app.navon.bike.domain.RoutePackageVersion
import app.navon.bike.domain.RoutePlanRequest
import app.navon.bike.domain.RoutePreviewModel
import app.navon.bike.domain.RouteProvenance
import app.navon.bike.domain.RouteProviderId
import app.navon.bike.domain.RerouteContext
import app.navon.bike.domain.RouteSummary
import app.navon.bike.domain.RoutingProvider
import app.navon.bike.integration.sample.SampleRoutingAdapter
import org.json.JSONArray
import org.json.JSONObject

/**
 * OSM cycling routing orchestrator. Issues parallel requests to BRouter
 * (`fastbike` paths-preferred + `trekking` balanced) AND OSRM bike, then
 * exposes whichever succeed as 1-3 alternatives in the route preview.
 *
 * Mirrors the web companion's `OsmCyclingRoutingAdapter.ts` byte-for-byte
 * in semantics. See `docs/companion-app-architecture.md` for the design.
 */
class OsmCyclingRoutingAdapter : RoutingProvider {
    companion object {
        private const val MIN_HEADING_SPEED_MPS = 2.0
        private const val REROUTE_FORWARD_SHIFT_M = 15.0
    }
    override val providerId: RouteProviderId = RouteProviderId.OSM
    override val isAvailableInV1: Boolean = true

    /** Last-resort fallback when all live sources fail. */
    private val sample = SampleRoutingAdapter(RouteProviderId.OSM)

    override suspend fun planRoute(request: RoutePlanRequest): RoutePreviewModel {
        return fanOutOrFallback(request, revision = 1)
    }

    override suspend fun replanRoute(
        session: ActiveRouteSession,
        riderLocation: CoordinatePoint,
        rerouteContext: RerouteContext?,
    ): RoutePreviewModel {
        val rerouteOrigin = headingBiasedOrigin(riderLocation, rerouteContext, "osm")
        val request = RoutePlanRequest(
            origin = rerouteOrigin,
            destination = session.destinationCoordinate ?: riderLocation,
            providerId = providerId,
        )
        return fanOutOrFallback(request, (session.routeRevision ?: 0) + 1)
    }

    private fun headingBiasedOrigin(
        riderLocation: CoordinatePoint,
        rerouteContext: RerouteContext?,
        providerLabel: String,
    ): CoordinatePoint {
        val heading = rerouteContext?.headingDegrees
        val speed = rerouteContext?.speedMps
        if (heading == null || !heading.isFinite()) {
            println("[reroute_heading] provider=$providerLabel reason=no_heading")
            return riderLocation
        }
        if (speed == null || !speed.isFinite() || speed < MIN_HEADING_SPEED_MPS) {
            println("[reroute_heading] provider=$providerLabel reason=low_speed speed=${speed ?: "nil"}")
            return riderLocation
        }
        val shifted = shiftPointByHeading(riderLocation, heading, REROUTE_FORWARD_SHIFT_M)
        if (!shifted.latitude.isFinite() || !shifted.longitude.isFinite() ||
            (shifted.latitude == riderLocation.latitude && shifted.longitude == riderLocation.longitude)
        ) {
            println("[reroute_heading] provider=$providerLabel reason=shift_failed")
            return riderLocation
        }
        println("[reroute_heading] provider=$providerLabel reason=applied")
        return shifted
    }

    private fun shiftPointByHeading(
        point: CoordinatePoint,
        headingDegrees: Double,
        distanceMeters: Double,
    ): CoordinatePoint {
        val metersPerDegLat = 111_320.0
        val rad = Math.toRadians(headingDegrees)
        val northM = kotlin.math.cos(rad) * distanceMeters
        val eastM = kotlin.math.sin(rad) * distanceMeters
        val lat = point.latitude + northM / metersPerDegLat
        val lonScale = metersPerDegLat * kotlin.math.cos(Math.toRadians(point.latitude))
        val lon = if (lonScale == 0.0) point.longitude else point.longitude + eastM / lonScale
        return CoordinatePoint(latitude = lat, longitude = lon)
    }

    override fun normalizePreview(
        preview: RoutePreviewModel,
        request: RoutePlanRequest,
    ): NormalizedRoutePackage {
        return preview.selectedAlternative?.normalizedPackage
            ?: error("OsmCyclingRoutingAdapter: preview has no alternatives")
    }

    private suspend fun fanOutOrFallback(
        request: RoutePlanRequest,
        revision: Int,
    ): RoutePreviewModel = coroutineScope {
        val tasks = listOf(
            async(Dispatchers.IO) {
                runCatching {
                    val feature = BrouterClient.fetch(BrouterProfile.FASTBIKE, request.origin, request.destination)
                    mapBrouterToAlternative(feature, request, revision, BrouterProfile.FASTBIKE, "Bike-paths first")
                }.getOrNull()
            },
            async(Dispatchers.IO) {
                runCatching {
                    val feature = BrouterClient.fetch(BrouterProfile.TREKKING, request.origin, request.destination)
                    mapBrouterToAlternative(feature, request, revision, BrouterProfile.TREKKING, "Balanced cycling")
                }.getOrNull()
            },
            async(Dispatchers.IO) {
                runCatching { fetchOsrmAlternative(request, revision) }.getOrNull()
            },
        )
        val results = tasks.map { it.await() }.filterNotNull()
        val deduped = dedupeAlternatives(results)
        if (deduped.isEmpty()) {
            return@coroutineScope sample.planRoute(request).copy(
                planningNotice = "Showing sample route — live routing failed",
            )
        }
        val failedCount = tasks.size - deduped.size
        RoutePreviewModel(
            alternatives = deduped,
            selectedAlternativeId = deduped.first().id,
            routeIdentifier = deduped.first().normalizedPackage.routeIdentifier,
            routeRevision = deduped.first().normalizedPackage.revision,
            planningNotice = if (failedCount == 0) {
                "Cycling alternatives via BRouter + OSRM"
            } else {
                "Cycling alternatives — $failedCount source${if (failedCount == 1) "" else "s"} unavailable"
            },
        )
    }

    private suspend fun fetchOsrmAlternative(
        request: RoutePlanRequest,
        revision: Int,
    ): RouteAlternative? = withContext(Dispatchers.IO) {
        val coordinates = String.format(
            Locale.US,
            "%.6f,%.6f;%.6f,%.6f",
            request.origin.longitude,
            request.origin.latitude,
            request.destination.longitude,
            request.destination.latitude,
        )
        val url = URL(
            "https://router.project-osrm.org/route/v1/bike/$coordinates?alternatives=false&overview=full&steps=true&geometries=geojson",
        )
        val connection = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 10_000
            readTimeout = 10_000
        }
        val statusCode = connection.responseCode
        val body = runCatching {
            val stream = if (statusCode in 200..299) connection.inputStream else connection.errorStream
            stream?.bufferedReader()?.use { it.readText() }.orEmpty()
        }.getOrDefault("")
        if (statusCode !in 200..299) throw IllegalStateException("HTTP $statusCode")
        val root = JSONObject(body)
        if (root.optString("code") != "Ok") throw IllegalStateException("OSRM error")
        val routes = root.optJSONArray("routes") ?: throw IllegalStateException("OSRM no routes")
        if (routes.length() == 0) throw IllegalStateException("OSRM empty routes")
        mapOsrmRoute(routes.getJSONObject(0), request, revision)
    }
}

private fun mapOsrmRoute(
    routeJson: JSONObject,
    request: RoutePlanRequest,
    revision: Int,
): RouteAlternative? {
    val geomJson = routeJson.optJSONObject("geometry") ?: return null
    val coordsJson = geomJson.optJSONArray("coordinates") ?: return null
    if (coordsJson.length() < 2) return null
    val geometry = mutableListOf<CoordinatePoint>()
    for (i in 0 until coordsJson.length()) {
        val pair = coordsJson.optJSONArray(i) ?: continue
        if (pair.length() < 2) continue
        geometry += CoordinatePoint(latitude = pair.optDouble(1), longitude = pair.optDouble(0))
    }
    if (geometry.size < 2) return null
    val distance = routeJson.optDouble("distance", 0.0)
    val duration = routeJson.optDouble("duration", 0.0).toInt().coerceAtLeast(60)
    val maneuvers = buildOsrmManeuvers(routeJson, geometry)
    val routePackage = NormalizedRoutePackage(
        version = RoutePackageVersion.CURRENT,
        routeIdentifier = osrmRouteId(request),
        revision = revision,
        geometry = geometry,
        maneuvers = maneuvers,
        summary = RouteSummary(
            totalDistanceMeters = distance,
            estimatedDurationSeconds = duration,
            startLabel = "Current location",
            destinationLabel = "Selected destination",
        ),
        provenance = RouteProvenance(
            providerId = RouteProviderId.OSM,
            sourceReference = "OSRM bike",
            generatedAtUnixMs = System.currentTimeMillis(),
        ),
    )
    val km = String.format(Locale.US, "%.1f", distance / 1000.0)
    val min = maxOf(duration / 60, 1)
    return RouteAlternative(
        id = UUID.randomUUID().toString(),
        title = "Fastest",
        subtitle = "$km km • $min min",
        distanceMeters = distance.toInt(),
        durationSeconds = duration,
        normalizedPackage = routePackage,
    )
}

private fun buildOsrmManeuvers(
    routeJson: JSONObject,
    geometry: List<CoordinatePoint>,
): List<RouteManeuver> {
    val maneuvers = mutableListOf<RouteManeuver>()
    val legsJson = routeJson.optJSONArray("legs") ?: JSONArray()
    val firstNext = firstOsrmStepDistance(legsJson)
    maneuvers += RouteManeuver(
        id = "depart",
        maneuverType = RouteManeuverType.DEPART,
        location = geometry.firstOrNull() ?: CoordinatePoint(0.0, 0.0),
        distanceFromStartMeters = 0.0,
        distanceToNextMeters = firstNext,
        instructionText = "Start riding",
    )
    var distanceFromStart = 0.0
    for (li in 0 until legsJson.length()) {
        val steps = legsJson.optJSONObject(li)?.optJSONArray("steps") ?: continue
        for (si in 0 until steps.length()) {
            val step = steps.getJSONObject(si)
            val distance = step.optDouble("distance", 0.0)
            val maneuverJson = step.optJSONObject("maneuver")
            if (maneuverJson == null) {
                distanceFromStart += distance
                continue
            }
            val type = maneuverJson.optString("type").lowercase(Locale.ROOT)
            if (type == "depart" || type == "arrive" || type == "notification" ||
                type == "new name" || type == "continue"
            ) {
                distanceFromStart += distance
                continue
            }
            val locationJson = maneuverJson.optJSONArray("location")
            if (locationJson == null || locationJson.length() < 2) {
                distanceFromStart += distance
                continue
            }
            val location = CoordinatePoint(
                latitude = locationJson.optDouble(1),
                longitude = locationJson.optDouble(0),
            )
            val name = step.optString("name")
            maneuvers += RouteManeuver(
                id = if (name.isBlank()) "step-${maneuvers.size}" else "step-${maneuvers.size}-$name",
                maneuverType = osrmManeuverType(type, maneuverJson.optString("modifier")),
                location = location,
                distanceFromStartMeters = distanceFromStart,
                distanceToNextMeters = distance.takeIf { it > 0.0 },
                instructionText = osrmInstructionText(type, maneuverJson.optString("modifier"), name),
            )
            distanceFromStart += distance
        }
    }
    maneuvers += RouteManeuver(
        id = "arrive",
        maneuverType = RouteManeuverType.ARRIVE,
        location = geometry.lastOrNull() ?: CoordinatePoint(0.0, 0.0),
        distanceFromStartMeters = routeJson.optDouble("distance", distanceFromStart),
        distanceToNextMeters = null,
        instructionText = "Arrive at destination",
    )
    return maneuvers
}

private fun firstOsrmStepDistance(legsJson: JSONArray): Double? {
    for (li in 0 until legsJson.length()) {
        val steps = legsJson.optJSONObject(li)?.optJSONArray("steps") ?: continue
        for (si in 0 until steps.length()) {
            val step = steps.getJSONObject(si)
            val type = step.optJSONObject("maneuver")?.optString("type")?.lowercase(Locale.ROOT)
                ?: continue
            if (type != "depart") return step.optDouble("distance").takeIf { it > 0.0 }
        }
    }
    return null
}

private fun osrmManeuverType(type: String, modifier: String?): RouteManeuverType = when (type) {
    "roundabout", "rotary" -> RouteManeuverType.ROUNDABOUT
    "merge", "fork", "on ramp", "off ramp" -> RouteManeuverType.MERGE
    "arrive" -> RouteManeuverType.ARRIVE
    else -> when (modifier?.lowercase(Locale.ROOT)) {
        "uturn" -> RouteManeuverType.UTURN
        "sharp right" -> RouteManeuverType.SHARP_RIGHT
        "right" -> RouteManeuverType.RIGHT
        "slight right" -> RouteManeuverType.SLIGHT_RIGHT
        "sharp left" -> RouteManeuverType.SHARP_LEFT
        "left" -> RouteManeuverType.LEFT
        "slight left" -> RouteManeuverType.SLIGHT_LEFT
        else -> RouteManeuverType.STRAIGHT
    }
}

private fun osrmInstructionText(type: String, modifier: String?, name: String): String = when (type) {
    "roundabout", "rotary" -> "Enter roundabout"
    "merge" -> "Merge"
    "fork" -> when {
        modifier?.contains("left", ignoreCase = true) == true -> "Keep left"
        modifier?.contains("right", ignoreCase = true) == true -> "Keep right"
        else -> "Keep to the fork"
    }
    else -> when (modifier?.lowercase(Locale.ROOT)) {
        "uturn" -> "Make a U-turn"
        "sharp right" -> "Turn sharply right"
        "right" -> "Turn right"
        "slight right" -> "Bear right"
        "sharp left" -> "Turn sharply left"
        "left" -> "Turn left"
        "slight left" -> "Bear left"
        else -> if (name.isBlank()) "Continue" else "Continue on $name"
    }
}

private fun osrmRouteId(request: RoutePlanRequest): String {
    val o = String.format(Locale.US, "%.5f,%.5f", request.origin.latitude, request.origin.longitude)
    val d = String.format(Locale.US, "%.5f,%.5f", request.destination.latitude, request.destination.longitude)
    return "osm-osrm:$o->$d"
}

private fun dedupeAlternatives(candidates: List<RouteAlternative>): List<RouteAlternative> {
    val kept = mutableListOf<RouteAlternative>()
    for (c in candidates) {
        if (kept.none { areNearIdentical(it, c) }) kept += c
    }
    return kept
}

private fun areNearIdentical(a: RouteAlternative, b: RouteAlternative): Boolean {
    val aLen = a.normalizedPackage.summary.totalDistanceMeters
    val bLen = b.normalizedPackage.summary.totalDistanceMeters
    if (aLen <= 0 || bLen <= 0) return false
    val lengthDelta = abs(aLen - bLen) / maxOf(aLen, bLen)
    if (lengthDelta > 0.03) return false
    val ag = a.normalizedPackage.geometry
    val bg = b.normalizedPackage.geometry
    if (ag.isEmpty() || bg.isEmpty()) return false
    return samePoint(ag.first(), bg.first()) &&
        samePoint(ag.last(), bg.last()) &&
        samePoint(ag[ag.size / 2], bg[bg.size / 2])
}

private fun samePoint(a: CoordinatePoint, b: CoordinatePoint): Boolean {
    return abs(a.latitude - b.latitude) < 1e-4 && abs(a.longitude - b.longitude) < 1e-4
}

File: companion-apps/ios/CompanionApp/Integration/Cycling/OsmCyclingRoutingAdapter.swift
import Foundation

/// Fans out to BRouter (fastbike + trekking) and OSRM in parallel, exposing whichever succeed as alternatives.
struct OsmCyclingRoutingAdapter: RoutingProvider {
    let providerID: RouteProviderID = .osm
    let isAvailableInV1: Bool = true
    private static let minHeadingSpeedMps: Double = 2.0
    private static let rerouteForwardShiftM: Double = 15.0

    /// Last-resort fallback when all live sources fail.
    private let sample = SampleRoutingAdapter(providerID: .osm)

    func planRoute(_ request: RoutePlanRequest) async throws -> RoutePreviewModel {
        return await fanOutOrFallback(request: request, revision: 1)
    }

    func replanRoute(
        using session: ActiveRouteSession,
        riderLocation: CoordinatePoint,
        rerouteContext: RerouteContext?
    ) async throws -> RoutePreviewModel {
        let rerouteOrigin = headingBiasedOrigin(
            riderLocation: riderLocation,
            rerouteContext: rerouteContext,
            providerLabel: "osm"
        )
        let request = RoutePlanRequest(
            origin: rerouteOrigin,
            destination: session.destinationCoordinate ?? riderLocation,
            providerID: providerID
        )
        return await fanOutOrFallback(request: request, revision: (session.routeRevision ?? 0) + 1)
    }

    func normalizePreview(_ preview: RoutePreviewModel, request: RoutePlanRequest) throws -> NormalizedRoutePackage {
        guard let selected = preview.selectedAlternative else {
            throw NSError(domain: "OsmCycling", code: 1, userInfo: [NSLocalizedDescriptionKey: "no alternatives"])
        }
        return selected.normalizedPackage
    }

    private func fanOutOrFallback(request: RoutePlanRequest, revision: Int) async -> RoutePreviewModel {
        async let fastbike = mapBrouter(request: request, revision: revision, profile: .fastbike, title: "Bike-paths first")
        async let trekking = mapBrouter(request: request, revision: revision, profile: .trekking, title: "Balanced cycling")
        async let osrm = fetchOsrmAlternative(request: request, revision: revision)
        let results: [RouteAlternative?] = [await fastbike, await trekking, await osrm]
        let successes = results.compactMap { $0 }
        let deduped = dedupeAlternatives(successes)
        if deduped.isEmpty {
            let fallback = (try? await sample.planRoute(request)) ?? RoutePreviewModel(
                alternatives: [], selectedAlternativeID: nil,
                routeIdentifier: nil, routeRevision: nil,
                planningNotice: T.string("cycling.sampleFallback")
            )
            return RoutePreviewModel(
                alternatives: fallback.alternatives,
                selectedAlternativeID: fallback.selectedAlternativeID,
                routeIdentifier: fallback.routeIdentifier,
                routeRevision: fallback.routeRevision,
                planningNotice: T.string("cycling.sampleFallback")
            )
        }
        let failedCount = results.count - deduped.count
        let notice = failedCount == 0
            ? T.string("cycling.alternativesLive")
            : T.string("cycling.alternativesPartial", ["count": .number(Double(failedCount))])
        return RoutePreviewModel(
            alternatives: deduped,
            selectedAlternativeID: deduped.first?.id,
            routeIdentifier: deduped.first?.normalizedPackage.routeIdentifier,
            routeRevision: deduped.first?.normalizedPackage.revision,
            planningNotice: notice
        )
    }

    private func mapBrouter(
        request: RoutePlanRequest,
        revision: Int,
        profile: BrouterProfile,
        title: String
    ) async -> RouteAlternative? {
        do {
            let feature = try await BrouterClient.fetch(profile: profile, origin: request.origin, destination: request.destination)
            return mapBrouterToAlternative(feature: feature, request: request, revision: revision, profile: profile, title: title)
        } catch {
            return nil
        }
    }

    private func fetchOsrmAlternative(request: RoutePlanRequest, revision: Int) async -> RouteAlternative? {
        let coordinates = String(
            format: "%.6f,%.6f;%.6f,%.6f",
            request.origin.longitude, request.origin.latitude,
            request.destination.longitude, request.destination.latitude
        )
        guard let url = URL(string: "https://router.project-osrm.org/route/v1/bike/\(coordinates)?alternatives=false&overview=full&steps=true&geometries=geojson") else {
            return nil
        }
        do {
            let (data, response) = try await URLSession.shared.data(from: url)
            guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
                return nil
            }
            guard let root = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  (root["code"] as? String) == "Ok",
                  let routes = root["routes"] as? [[String: Any]],
                  let first = routes.first else {
                return nil
            }
            return mapOsrmRoute(routeJson: first, request: request, revision: revision)
        } catch {
            return nil
        }
    }

    private func headingBiasedOrigin(
        riderLocation: CoordinatePoint,
        rerouteContext: RerouteContext?,
        providerLabel: String
    ) -> CoordinatePoint {
        guard let heading = rerouteContext?.headingDegrees, heading.isFinite else {
            return riderLocation
        }
        guard let speed = rerouteContext?.speedMps, speed.isFinite, speed >= Self.minHeadingSpeedMps else {
            return riderLocation
        }
        let shifted = shiftPointByHeading(point: riderLocation, headingDegrees: heading, distanceMeters: Self.rerouteForwardShiftM)
        if !shifted.latitude.isFinite || !shifted.longitude.isFinite ||
            (shifted.latitude == riderLocation.latitude && shifted.longitude == riderLocation.longitude) {
            return riderLocation
        }
        return shifted
    }

    private func shiftPointByHeading(
        point: CoordinatePoint,
        headingDegrees: Double,
        distanceMeters: Double
    ) -> CoordinatePoint {
        let metersPerDegLat = 111_320.0
        let rad = headingDegrees * .pi / 180.0
        let northM = cos(rad) * distanceMeters
        let eastM = sin(rad) * distanceMeters
        let latitude = point.latitude + northM / metersPerDegLat
        let lonScale = metersPerDegLat * cos(point.latitude * .pi / 180.0)
        let longitude = lonScale == 0
            ? point.longitude
            : point.longitude + eastM / lonScale
        return CoordinatePoint(latitude: latitude, longitude: longitude)
    }
}

private func mapOsrmRoute(routeJson: [String: Any], request: RoutePlanRequest, revision: Int) -> RouteAlternative? {
    guard let geom = routeJson["geometry"] as? [String: Any],
          let coordsRaw = geom["coordinates"] as? [[Double]] else { return nil }
    let geometry: [CoordinatePoint] = coordsRaw.compactMap { c in
        guard c.count >= 2 else { return nil }
        return CoordinatePoint(latitude: c[1], longitude: c[0])
    }
    guard geometry.count >= 2 else { return nil }
    let distance = (routeJson["distance"] as? Double) ?? 0.0
    let duration = max(Int((routeJson["duration"] as? Double) ?? 0), 60)
    let maneuvers = buildOsrmManeuvers(routeJson: routeJson, geometry: geometry)
    let routePackage = NormalizedRoutePackage(
        version: RoutePackageVersion.current,
        routeIdentifier: osrmRouteId(request: request),
        revision: revision,
        geometry: geometry,
        maneuvers: maneuvers,
        summary: RouteSummary(
            totalDistanceMeters: distance,
            estimatedDurationSeconds: duration,
            startLabel: "Current location",
            destinationLabel: "Selected destination"
        ),
        provenance: RouteProvenance(
            providerID: .osm,
            sourceReference: "OSRM bike",
            generatedAtUnixMs: UInt64(Date().timeIntervalSince1970 * 1000)
        )
    )
    let km = String(format: "%.1f", distance / 1000.0)
    let min = max(duration / 60, 1)
    return RouteAlternative(
        id: UUID(),
        title: "Fastest",
        subtitle: "\(km) km • \(min) min",
        distanceMeters: Int(distance),
        durationSeconds: duration,
        normalizedPackage: routePackage
    )
}

private func buildOsrmManeuvers(routeJson: [String: Any], geometry: [CoordinatePoint]) -> [RouteManeuver] {
    var maneuvers: [RouteManeuver] = []
    let legs = routeJson["legs"] as? [[String: Any]] ?? []
    let firstNext = firstOsrmStepDistance(legs: legs)
    maneuvers.append(RouteManeuver(
        id: "depart", maneuverType: .depart,
        location: geometry.first ?? CoordinatePoint(latitude: 0, longitude: 0),
        distanceFromStartMeters: 0, distanceToNextMeters: firstNext,
        instructionText: "Start riding"
    ))
    var distanceFromStart: Double = 0.0
    for leg in legs {
        let steps = leg["steps"] as? [[String: Any]] ?? []
        for step in steps {
            let distance = (step["distance"] as? Double) ?? 0.0
            guard let m = step["maneuver"] as? [String: Any] else {
                distanceFromStart += distance
                continue
            }
            let type = ((m["type"] as? String) ?? "").lowercased()
            if type == "depart" || type == "arrive" || type == "notification" || type == "new name" || type == "continue" {
                distanceFromStart += distance
                continue
            }
            guard let location = m["location"] as? [Double], location.count >= 2 else {
                distanceFromStart += distance
                continue
            }
            let modifier = (m["modifier"] as? String) ?? ""
            let name = (step["name"] as? String) ?? ""
            maneuvers.append(RouteManeuver(
                id: name.isEmpty ? "step-\(maneuvers.count)" : "step-\(maneuvers.count)-\(name)",
                maneuverType: osrmManeuverType(type: type, modifier: modifier),
                location: CoordinatePoint(latitude: location[1], longitude: location[0]),
                distanceFromStartMeters: distanceFromStart,
                distanceToNextMeters: distance > 0 ? distance : nil,
                instructionText: osrmInstruction(type: type, modifier: modifier, name: name)
            ))
            distanceFromStart += distance
        }
    }
    maneuvers.append(RouteManeuver(
        id: "arrive", maneuverType: .arrive,
        location: geometry.last ?? CoordinatePoint(latitude: 0, longitude: 0),
        distanceFromStartMeters: (routeJson["distance"] as? Double) ?? distanceFromStart,
        distanceToNextMeters: nil,
        instructionText: "Arrive at destination"
    ))
    return maneuvers
}

private func firstOsrmStepDistance(legs: [[String: Any]]) -> Double? {
    for leg in legs {
        let steps = leg["steps"] as? [[String: Any]] ?? []
        for step in steps {
            let m = step["maneuver"] as? [String: Any]
            let type = ((m?["type"] as? String) ?? "").lowercased()
            if type != "depart" {
                if let d = step["distance"] as? Double, d > 0 { return d }
            }
        }
    }
    return nil
}

private func osrmManeuverType(type: String, modifier: String) -> RouteManeuverType {
    switch type {
    case "roundabout", "rotary": return .roundabout
    case "merge", "fork", "on ramp", "off ramp": return .merge
    case "arrive": return .arrive
    default:
        switch modifier.lowercased() {
        case "uturn": return .uturn
        case "sharp right": return .sharpRight
        case "right": return .right
        case "slight right": return .slightRight
        case "sharp left": return .sharpLeft
        case "left": return .left
        case "slight left": return .slightLeft
        default: return .straight
        }
    }
}

private func osrmInstruction(type: String, modifier: String, name: String) -> String {
    switch type {
    case "roundabout", "rotary": return "Enter roundabout"
    case "merge": return "Merge"
    case "fork":
        if modifier.lowercased().contains("left") { return "Keep left" }
        if modifier.lowercased().contains("right") { return "Keep right" }
        return "Keep to the fork"
    default:
        switch modifier.lowercased() {
        case "uturn": return "Make a U-turn"
        case "sharp right": return "Turn sharply right"
        case "right": return "Turn right"
        case "slight right": return "Slight right"
        case "sharp left": return "Turn sharply left"
        case "left": return "Turn left"
        case "slight left": return "Slight left"
        default: return name.isEmpty ? "Continue" : "Continue on \(name)"
        }
    }
}

private func osrmRouteId(request: RoutePlanRequest) -> String {
    let o = String(format: "%.5f,%.5f", request.origin.latitude, request.origin.longitude)
    let d = String(format: "%.5f,%.5f", request.destination.latitude, request.destination.longitude)
    return "osm-osrm:\(o)->\(d)"
}

private func dedupeAlternatives(_ candidates: [RouteAlternative]) -> [RouteAlternative] {
    var kept: [RouteAlternative] = []
    for c in candidates where !kept.contains(where: { areNearIdentical($0, c) }) {
        kept.append(c)
    }
    return kept
}

private func areNearIdentical(_ a: RouteAlternative, _ b: RouteAlternative) -> Bool {
    let aLen = a.normalizedPackage.summary.totalDistanceMeters
    let bLen = b.normalizedPackage.summary.totalDistanceMeters
    if aLen <= 0 || bLen <= 0 { return false }
    let lengthDelta = abs(aLen - bLen) / max(aLen, bLen)
    if lengthDelta > 0.03 { return false }
    let ag = a.normalizedPackage.geometry
    let bg = b.normalizedPackage.geometry
    if ag.isEmpty || bg.isEmpty { return false }
    return samePoint(ag.first!, bg.first!) &&
        samePoint(ag.last!, bg.last!) &&
        samePoint(ag[ag.count / 2], bg[bg.count / 2])
}

private func samePoint(_ a: CoordinatePoint, _ b: CoordinatePoint) -> Bool {
    return abs(a.latitude - b.latitude) < 1e-4 && abs(a.longitude - b.longitude) < 1e-4
}
